cw/util/trace_library.py

```diff
@@ -0,0 +1,194 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+import datetime
+from dataclasses import asdict, dataclass
+from typing import Optional
+
+import numpy as np
+import sqlalchemy as db
+import sqlalchemy_utils as db_utils
+from sqlalchemy.orm import sessionmaker
+
+
+@dataclass
+class Metadata:
+    """ Metadata for capture.
+
+    Stores information about the capture of the traces.
+
+    """
+    config: str
+    datetime: datetime.datetime
+    bitstream: str
+    binary: str
+    offset: Optional[int] = 0
+    sample_rate: Optional[int] = 0
+    scope_gain: Optional[float] = 0
+    trigger_level: Optional[float] = 0
+    time_div: Optional[float] = 0
+
+
+@dataclass
+class Trace:
+    """ Trace.
+
+    Defines the trace format.
+
+    """
+    wave: bytearray
+    plaintext: bytearray
+    ciphertext: bytearray
+    key: bytearray
+    x_pos: Optional[int] = 0
+    y_pos: Optional[int] = 0
+    trace_id: Optional[int] = 0
+
+
+class TraceLibrary:
+    """ Class for writing/reading traces to a database.
+
+    Either a new database is created or traces are appended to the existing
+    one. Traces are first written into memory and the flushed into the
+    database after reaching a trace memory threshold.
+    """
+    def __init__(self, db_name, trace_threshold):
+        self.engine = db.create_engine("sqlite:///" + db_name + ".db")
+        db_utils.create_database(self.engine.url)
+        self.Session = sessionmaker(self.engine)
+        self.session = self.Session()
```

Comment:
Is there a need for having both capitalisations? If not, maybe it's easier to pick one?